jina/drivers/__init__.py

```diff
@@ -305,93 +320,41 @@ def __getstate__(self) -> Dict[str, Any]:
         return d
 
 
-class RecursiveMixin:
-    """A mixin to traverse a set of Documents with a specific path. to be mixed in with :class:`BaseRecursiveDriver`"""
-
-    @property
-    def docs(self):
-        """
+class ContextAwareRecursiveMixin:
+    """
+     The full datastructure version of :class:`FlatRecursiveMixin`, to be mixed in with :class:`BaseRecursiveDriver`.
+     It uses :meth:`traverse` in :class:`DocumentSet` and allows direct manipulation of Chunk-/Match-/DocumentSets.
 
+     .. seealso::
+        https://github.com/jina-ai/jina/issues/1932
 
-        .. # noqa: DAR102
+    """
 
+    def __call__(self, *args, **kwargs):
+        """Traverse with _apply_all
 
-        .. # noqa: DAR201
+        :param *args: *args forwarded to ``_apply_all``
+        :param **kwargs: **kwargs forwarded to ``_apply_all``
         """
-        if self.expect_parts > 1:
-            return (d for r in reversed(self.partial_reqs) for d in r.docs)
-        else:
-            return self.req.docs
-
-    def _apply_root(
-        self,
-        docs: 'DocumentSet',
-        field: str,
-        *args,
-        **kwargs,
-    ) -> None:
-        return self._apply_all(docs, None, field, *args, **kwargs)
+        document_sets = self.docs.traverse(self._traversal_paths)
+        self._apply_all(document_sets, *args, **kwargs)
 
-    # TODO(Han): probably want to publicize this, as it is not obvious for driver
-    #  developer which one should be inherited
     def _apply_all(
         self,
-        docs: 'DocumentSet',
-        context_doc: 'Document',
-        field: str,
+        doc_sequences: Iterable['DocumentSet'],
         *args,
         **kwargs,
     ) -> None:
-        """Apply function works on a list of docs, modify the docs in-place
-
-        :param docs: a list of :class:`jina.Document` objects to work on; they could come from ``matches``/``chunks``.
-        :param context_doc: the owner of ``docs``
-        :param field: where ``docs`` comes from, either ``matches`` or ``chunks``
-        :param *args: *args
-        :param **kwargs: **kwargs
-        """
-
-    def __call__(self, *args, **kwargs):
-        """Call the Driver.
+        """Apply function works on a list of list of docs, modify the docs in-place.
```

Comment:
Is it a list of docs instead of a list of list of docs?

Reply:
It is an `Iterable[DocumentSet]` which is an `Iterable` over Chunks or Matches. Witch are a list of Documents.

Reply:
then, we should change the docstring I guess:
`iterable over docsets`